// @vscode/wasm-component-model/lib/browser/connection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkerConnection = exports.MainConnection = void 0;
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */
const connection_1 = require("../common/connection");
class MainConnection extends connection_1.BaseMainConnection {
    port;
    constructor(port) {
        super();
        this.port = port;
    }
    dispose() {
        this.port.onmessage = null;
        if (this.port instanceof MessagePort) {
            this.port.close();
        }
        super.dispose();
    }
    postMessage(message, transfer) {
        transfer !== undefined ? this.port.postMessage(message, transfer) : this.port.postMessage(message);
    }
    listen() {
        this.port.onmessage = (event) => {
            this.handleMessage(event.data);
        };
    }
}
exports.MainConnection = MainConnection;
class WorkerConnection extends connection_1.BaseWorkerConnection {
    port;
    constructor(port, world, timeout) {
        super(world, timeout);
        this.port = port;
    }
    dispose() {
        this.port.onmessage = null;
        super.dispose();
    }
    postMessage(message, transfer) {
        transfer !== undefined ? this.port.postMessage(message, transfer) : this.port.postMessage(message);
    }
    listen() {
        this.port.onmessage = (event) => {
            this.handleMessage(event.data);
        };
    }
}
exports.WorkerConnection = WorkerConnection;
